src/dbnode/storage/index/wide_query_results_test.go

```diff
@@ -0,0 +1,208 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/m3db/m3/src/m3ninx/doc"
+	"github.com/m3db/m3/src/x/ident"
+	"github.com/m3db/m3/src/x/pool"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	testNs     = ident.StringID("ns")
+	testIDPool ident.Pool
+)
+
+func init() {
+	// NB: build test pools once.
+	bytesPool := pool.NewCheckedBytesPool(nil, nil,
+		func(s []pool.Bucket) pool.BytesPool {
+			return pool.NewBytesPool(s, nil)
+		})
+
+	testIDPool = ident.NewPool(bytesPool, ident.PoolOptions{})
+	bytesPool.Init()
+}
+
+func buildExpected(
+	t *testing.T,
+	elementCount int,
+	batchSize int,
+) [][]string {
+	exBatches := int(math.Ceil(float64(elementCount) / float64(batchSize)))
+	expected := make([][]string, 0, exBatches)
+	for i := 0; i < exBatches; i++ {
+		batch := make([]string, 0, batchSize)
+		for j := 0; j < batchSize; j++ {
+			val := i*batchSize + j
+			if val < elementCount {
+				batch = append(batch, fmt.Sprintf("foo%d", val))
+			}
+		}
+
+		expected = append(expected, batch)
+	}
+
+	return expected
+}
+
+func assertExpected(
+	t *testing.T,
+	expected [][]string,
+	batchCh <-chan *ident.IDBatch,
+	doneCh chan<- struct{},
+) {
+	go func() {
+		i := 0
+		for batch := range batchCh {
+			batchStr := make([]string, 0, len(batch.IDs))
+			for _, b := range batch.IDs {
+				batchStr = append(batchStr, b.String())
+				b.Finalize()
+			}
+
+			batch.Done()
+			withinIndex := i < len(expected)
+			assert.True(t, withinIndex)
+			if withinIndex {
+				assert.Equal(t, expected[i], batchStr)
+			}
+			i++
+		}
+		assert.Equal(t, len(expected), i)
+		doneCh <- struct{}{}
+	}()
+}
+
+func buildDocs(elementCount int, batchSize int) [][]doc.Document {
+	docBatches := int(math.Ceil(float64(elementCount) / float64(batchSize)))
+	docs := make([][]doc.Document, 0, docBatches)
+	for i := 0; i < docBatches; i++ {
+		batch := make([]doc.Document, 0, batchSize)
+		for j := 0; j < batchSize; j++ {
+			val := i*batchSize + j
+			if val < elementCount {
+				val := fmt.Sprintf("foo%d", i*batchSize+j)
+				batch = append(batch, doc.Document{
+					ID: []byte(val),
+				})
+			}
+		}
+
+		docs = append(docs, batch)
+	}
+
+	return docs
+}
+
+func TestWideSeriesResults(t *testing.T) {
+	// Test many different permutations of element count and batch sizes.
+	max := 31
+	// elementCount := 30
+	// docBatchSize := 20
+	// batchSize := 10
+	for elementCount := 0; elementCount < max; elementCount++ {
+		for docBatchSize := 1; docBatchSize < max; docBatchSize++ {
+			for batchSize := 1; batchSize < max; batchSize++ {
+				var (
+					batchCh = make(chan *ident.IDBatch)
+					doneCh  = make(chan struct{})
+				)
+
+				docs := buildDocs(elementCount, docBatchSize)
+				expected := buildExpected(t, elementCount, batchSize)
+				assertExpected(t, expected, batchCh, doneCh)
```

Comment:
This needs to happen before `wideRes.AddDocuments(docBatch)` since this kicks off the consumer channel; otherwise we'd hit deadlocks here. Renamed this instead